hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/compactions/DefaultCompactor.java

```diff
@@ -17,42 +17,45 @@
  */
 package org.apache.hadoop.hbase.regionserver.compactions;
 
+import static org.apache.hadoop.hbase.regionserver.DefaultStoreEngine.DEFAULT_COMPACTION_ENABLE_DUAL_FILE_WRITER_KEY;
+import static org.apache.hadoop.hbase.regionserver.DefaultStoreEngine.DEFAULT_ENABLE_DUAL_FILE_WRITER;
+
 import java.io.IOException;
 import java.util.List;
 import java.util.function.Consumer;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.regionserver.DualFileWriter;
 import org.apache.hadoop.hbase.regionserver.HStore;
 import org.apache.hadoop.hbase.regionserver.InternalScanner;
-import org.apache.hadoop.hbase.regionserver.StoreFileWriter;
 import org.apache.hadoop.hbase.regionserver.throttle.ThroughputController;
 import org.apache.hadoop.hbase.security.User;
 import org.apache.yetus.audience.InterfaceAudience;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import org.apache.hbase.thirdparty.com.google.common.collect.Lists;
 
 /**
  * Compact passed set of files. Create an instance and then call
  * {@link #compact(CompactionRequestImpl, ThroughputController, User)}
  */
 @InterfaceAudience.Private
-public class DefaultCompactor extends Compactor<StoreFileWriter> {
-  private static final Logger LOG = LoggerFactory.getLogger(DefaultCompactor.class);
+public class DefaultCompactor extends AbstractMultiOutputCompactor<DualFileWriter> {
 
   public DefaultCompactor(Configuration conf, HStore store) {
     super(conf, store);
   }
 
-  private final CellSinkFactory<StoreFileWriter> writerFactory =
-    new CellSinkFactory<StoreFileWriter>() {
+  private final CellSinkFactory<DualFileWriter> writerFactory =
+    new CellSinkFactory<DualFileWriter>() {
       @Override
-      public StoreFileWriter createWriter(InternalScanner scanner, FileDetails fd,
+      public DualFileWriter createWriter(InternalScanner scanner, FileDetails fd,
         boolean shouldDropBehind, boolean major, Consumer<Path> writerCreationTracker)
         throws IOException {
-        return DefaultCompactor.this.createWriter(fd, shouldDropBehind, major,
-          writerCreationTracker);
+        boolean enableDualFileWriter = conf.getBoolean(
```

Comment:
It seems a bit strange here, the flag is named enable dual file writer, but no matter it is true or false, we always create a DualFileWriter...

I think here we should still return a StoreFileWriter? If dual file writer is enabled, we return a DualFileWriter, otherwise we return the default store file writer?

Reply:
StoreFileWriter is a single file writer. DualFileWriter can write to up to two files. It can behaves as a single file writer as well as a dual file writer. DualFileWriter extends AbstractMultiFileWriter as DateTieredMultiFileWriter and StripeMultiFileWriter do. The relationship type between StoreFileWriter and an AbstractMultiFileWriter is composition but not inheritance. Thus, they cannot be exchangeable. The store engine code is organized this way currently. Changing this organization would be very invasive.

Reply:
As said above, then we should just change the StoreFileWriter, not introducing a new DualFileWriter but can still write one file...